Change recieving options in Itamae::Resource::Base at initialize time

ganmacs opened the pull request:
ref #136 

When `dry_run` mode, raise an error in `edit` action against a `file` resource if the target file does not exist.
So, I skip the method that raise an error to use `dry_run` options.

this is recipe.rb
```rb
file "./sample_file" do
  action :edit
end
```

Before apply this patch
```sh
ganmacs@ganmacs~% bundle exec bin/itamae ssh --vagrant recipe.rb --dry-run
 INFO : Starting Itamae...
 INFO : Recipe: /Users/ganmacs/src/github.com/ganmacs/itamae/recipe.rb
/Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-scp-1.2.1/lib/net/scp.rb:365:in `block (3 levels) in start_command': SCP did not finish successfully (1): scp: ./sample_file: No such file or directory (Net::SCP::Error)
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/channel.rb:591:in `call'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/channel.rb:591:in `do_close'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:587:in `channel_close'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:466:in `dispatch_incoming_packets'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:222:in `preprocess'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:206:in `process'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:170:in `block in loop'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:170:in `loop'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/session.rb:170:in `loop'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-ssh-2.9.2/lib/net/ssh/connection/channel.rb:269:in `wait'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/net-scp-1.2.1/lib/net/scp.rb:321:in `download!'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/ext/specinfra.rb:34:in `scp_download!'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/ext/specinfra.rb:23:in `receive_file'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/backend.rb:93:in `receive_file'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/file.rb:23:in `pre_action'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:179:in `block in run_action'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/logger.rb:40:in `call'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/logger.rb:40:in `with_indent_if'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:177:in `run_action'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:135:in `block (2 levels) in run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:134:in `each'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:134:in `block in run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/logger.rb:40:in `call'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/logger.rb:40:in `with_indent_if'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/resource/base.rb:125:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:60:in `block in run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:57:in `each'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:57:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe.rb:48:in `block in run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/logger.rb:31:in `with_indent'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe.rb:47:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:62:in `block in run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:57:in `each'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/recipe_children.rb:57:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/runner.rb:53:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/runner.rb:23:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/lib/itamae/cli.rb:53:in `ssh'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /Users/ganmacs/src/github.com/ganmacs/itamae/vendor/bundle/ruby/2.2.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from bin/itamae:4:in `<main>'
bundle exec bin/itamae ssh --vagrant recipe.rb --dry-run  5.10s user 0.95s system 92% cpu 6.558 total
```


After apply this patch
```sh
anmacs@ganmacs~% bundle exec bin/itamae ssh --vagrant recipe.rb --dry-run
 INFO : Starting Itamae...
 INFO : Recipe: /Users/ganmacs/src/github.com/ganmacs/itamae/recipe.rb
 INFO :   file[./sample_file] exist will change from 'false' to 'true'
bundle exec bin/itamae ssh --vagrant recipe.rb --dry-run  5.08s user 0.94s system 89% cpu 6.738 total
```

As above, after Apply this patch, if the target file does not exist, we can't detect an error In `dry_run`

ryotarai commented on lib/itamae/resource/file.rb:
```diff
@@ -20,12 +20,16 @@ def pre_action
         end
 
-        send_tempfile
+        unless options[:dry_run]
```
`create` action in dry-run mode shows difference between tempfile and current file. This break the ability.

ryotarai commented:
Changing `pre_action` to accept an arg introduces plugin incompatibility because some plugin resources override `pre_action` method. I prefer:

```ruby
# base.rb
if method(:pre_action).arity == 1
  pre_action(options)
else
  Logger.warn "DEPRECATION WARNING: #{resource_type} resource does not accept `options` arg."
  pre_action
end

# resource
# option arg is not optional
def pre_action(options)
  # ...
end
```

ganmacs commented:
Thanks for a detail explanation.
I had not noticed that my changing make plugins incompatible.

I think receiving options as arguments in `Itamae::Resource::Base` object at initialize time can solve #136 and have compatible other plugins.
So I changed my patch drastically. and changed my PR title.

ryotarai commented:
Sorry for late reply. I'll check this PR later.

ryotarai commented:
Sorry for late reaction. Your commits are merged:

- https://github.com/itamae-kitchen/itamae/commit/54894fda544bc7fc38cc40d098af44e7757ebf0c
- https://github.com/itamae-kitchen/itamae/commit/0c57df5fc97aacce3b573e6f234e4232cf25c825

Thank you for your contribution!
